core-backend/wasmtime/src/env.rs

```diff
@@ -18,269 +18,162 @@
 
 //! Wasmtime environment for running a module.
 
+use alloc::rc::Rc;
+
 use crate::memory::MemoryWrap;
 use alloc::{boxed::Box, collections::BTreeMap, format, string::ToString, vec::Vec};
 use gear_backend_common::{
-    funcs, BackendError, BackendReport, Environment, ExtInfo, TerminationReason,
+    funcs as common_funcs, BackendError, BackendReport, Environment, ExtInfo, TerminationReason,
 };
 use gear_core::{
     env::{Ext, LaterExt},
     memory::{Memory, PageBuf, PageNumber},
 };
 use wasmtime::{
-    Extern, Func, Instance, Limits, Memory as WasmtimeMemory, MemoryType, Module, Store, Trap,
+    Engine, Extern, Func, Instance, Memory as WasmtimeMemory, MemoryType, Module, Store, Trap,
 };
 
+pub struct LaterStore<T>(Rc<Store<T>>);
+
+impl<T> Clone for LaterStore<T> {
+    fn clone(&self) -> Self {
+        LaterStore(self.0.clone())
+    }
+}
+
+impl<T: Default> LaterStore<T> {
+    pub fn new(eng: &Engine) -> Self {
+        LaterStore(Rc::from(Store::new(eng, T::default())))
+    }
+    /// In order to be able borrow mutable reference many times we need
+    /// to make it in unsafe manner.
+    /// Wasmtime store object must be mut borrowed to execute instance,
+    /// but also we must mut borrow it in memory sys-calls: alloc/free/...
+    /// But memory syscalls called in the same time with instance execution,
+    /// so there is no ways to avoid twice mut borrowing.
+    pub fn get_mut_ref(&mut self) -> &mut Store<T> {
+        unsafe {
+            let r = self.0.as_ref();
+            let ptr = r as *const Store<T> as *mut Store<T>;
+            ptr.as_mut().expect("ptr must be here")
+        }
+    }
+}
+
 /// Environment to run one module at a time providing Ext.
 pub struct WasmtimeEnvironment<E: Ext + 'static> {
-    store: wasmtime::Store,
+    store: LaterStore<()>,
```

Comment:
Why `LaterStore` is generic but used only with the unit type?

Reply:
I can remove it if you want. I add it as template, because LaterStore is just wrapper for Store, so should also be with T parameter. But it's not strict